rr_common/src/planner/global_path.cpp

```diff
@@ -0,0 +1,172 @@
+        for (auto path_point : global_segment) {
+            geometry_msgs::PoseStamped ps;
+            ps.pose.position.x = path_point.getX();
+            ps.pose.position.y = path_point.getY();
+            global_seg_msg.poses.push_back(ps);
+        }
+        global_seg_msg.header.frame_id = "map";
+        global_path_seg_pub_.publish(global_seg_msg);
+    }
+    return dtw_val;
+}
+
+std::vector<tf::Point> GlobalPath::get_global_segment(const std::vector<tf::Point> &sample_path) {
+    // get a list of each of the global points distances from the origin of the sample path
+    tf::Point sample_origin = sample_path[0];
+    std::vector<double> distances_to_origin(global_path_.size());
+    std::transform(global_path_.begin(), global_path_.end(), distances_to_origin.begin(),
+                   [sample_origin](const tf::Point &global_pnt) {
+                       return GlobalPath::GetPointDistance(global_pnt, sample_origin);
+                   });
+    // get the index of the closest global point to the start of the sample path
+    int seg_start_index =
+          std::min_element(distances_to_origin.begin(), distances_to_origin.end()) - distances_to_origin.begin();
+
+    // get the length of sample path by summing all the dists btwn the points
+    std::vector<double> sample_adj_dist = GlobalPath::adjacent_distances(sample_path);
+    double sample_length = std::accumulate(sample_adj_dist.begin(), sample_adj_dist.end(), 0.0);
+    // get the ending point of the global segment using the length of the segment
+    double upper_cum_limit =
+          std::fmod((global_cum_dist_[seg_start_index] + sample_length), global_cum_dist_[global_cum_dist_.size() - 1]);
+    int seg_end_index = std::upper_bound(global_cum_dist_.begin(), global_cum_dist_.end(), upper_cum_limit) -
+                        global_cum_dist_.begin();
+
+    // create the global segment
+    std::vector<tf::Point> global_segment;
+    if (seg_start_index <= seg_end_index) {  // assume sample path length < global path length
+        global_segment =
+              std::vector<tf::Point>(global_path_.begin() + seg_start_index, global_path_.begin() + seg_end_index);
+    } else {
+        global_segment = std::vector<tf::Point>(global_path_.begin() + seg_start_index, global_path_.end());
+        global_segment.insert(global_segment.end(), global_path_.begin(), global_path_.begin() + seg_end_index);
+    }
+    return global_segment;
+}
+
+double GlobalPath::dtw_distance(const std::vector<tf::Point> &path1, const std::vector<tf::Point> &path2, int w) {
+    int n = path1.size();
+    int m = path2.size();
+
+    std::vector<std::vector<double>> dtw(n + 1, std::vector<double>(m + 1, INFINITY));
+    w = std::max(w, std::abs(n - m));  // adapt window size
+    dtw[0][0] = 0;
+
+    for (int i = 1; i < n + 1; i++) {
+        for (int j = std::max(1, i - w); j < std::min(m, i + w) + 1; j++) {
+            dtw[i][j] = 0;
+        }
+    }
```

Comment:
I still think you don't need to put 0 for the entire part of the matrix we will be exploring.